google/cloud/security/common/gcp_api/_base_repository.py

```diff
@@ -0,0 +1,695 @@
+
+        if use_rate_limiter:
+            self._rate_limiter = RateLimiter(max_calls=quota_max_calls,
+                                             period=quota_period)
+        else:
+            self._rate_limiter = None
+
+        self.name = api_name
+
+        # Look to see if the API is formally supported in Forseti.
+        supported_api = _supported_apis.SUPPORTED_APIS.get(api_name)
+        if not supported_api:
+            LOGGER.warn('API "%s" is not formally supported in Forseti, '
+                        'proceed at your own risk.', api_name)
+
+        # See if the version is supported by Forseti.
+        # If no version is specified, use the supported API's default version.
+        if not versions and supported_api:
+            versions = [supported_api.get('default_version')]
+        self.versions = versions
+
+        if supported_api:
+            for version in versions:
+                if version not in supported_api.get('supported_versions', []):
+                    LOGGER.warn('API "%s" version %s is not formally supported '
+                                'in Forseti, proceed at your own risk.',
+                                api_name, version)
+
+        self.gcp_services = {}
+        for version in versions:
+            self.gcp_services[version] = _create_service_api(
+                self._credentials,
+                self.name,
+                version,
+                kwargs.get('developer_key'),
+                kwargs.get('cache_discovery', False))
+
+    def __repr__(self):
+        """The object representation.
+
+        Returns:
+            str: The object representation.
+        """
+        return 'API: name=%s, versions=%s' % (self.name, self.versions)
+
+    def _init_repository(self, repository_class, version=None):
+        """Safely initialize a repository class to a property.
+
+        Args:
+          repository_class (class): The class to initialize.
+          version (str): The gcp service version for the repository.
+
+        Returns:
+          object: An instance of repository_class.
+        """
+        if not version:
+            # Use either the default version if defined or the first version
+            # returned when sorted by name.
+            version = (
+                _supported_apis.SUPPORTED_APIS.get(self.name, {})
+                .get('default_version'))
+            if not version or version not in self.gcp_services:
+                version = sorted(self.gcp_services.keys())[0]
+
+        with self._repository_lock:
+            return repository_class(gcp_service=self.gcp_services[version],
+                                    credentials=self._credentials,
+                                    rate_limiter=self._rate_limiter)
+
+
+# pylint: disable=too-many-instance-attributes, too-many-arguments
+class GCPRepository(object):
+    """Base class for GCP APIs."""
+
+    def __init__(self, gcp_service, credentials, component, entity='',
```

Comment:
Is an empty string a reasonable default for entity?

Reply:
Some APIs do not use the entity target for Get, just another inconsistency.

Reply:
Yes, but I feel that the default should reflect the common case, which I believe to be not an empty string. In other words, it might be better to let the exceptions supply an empty string.
At any rate, in the future, we could detect such edge cases via inspection. Is Projects the only edge case at the moment?